File: src/main/scala/li/cil/oc/common/capabilities/CapabilityColored.scala
package li.cil.oc.common.capabilities

import li.cil.oc.api.internal.Colored
import li.cil.oc.integration.Mods
import net.minecraft.nbt.INBT
import net.minecraft.nbt.IntNBT
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.Direction
import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.ICapabilityProvider
import net.minecraftforge.common.util.LazyOptional
import net.minecraftforge.common.util.NonNullSupplier

object CapabilityColored {
  final val ProviderColored = new ResourceLocation(Mods.IDs.OpenComputers, "colored")

  class Provider(val tileEntity: TileEntity with Colored) extends ICapabilityProvider with NonNullSupplier[Provider] with Colored {
    private val wrapper = LazyOptional.of(this)

    def get = this

    def invalidate() = wrapper.invalidate

    override def getCapability[T](capability: Capability[T], facing: Direction): LazyOptional[T] = {
      if (capability == Capabilities.ColoredCapability) wrapper.cast[T]
      else LazyOptional.empty[T]
    }

    override def getColor = tileEntity.getColor

    override def setColor(value: Int) = tileEntity.setColor(value)

    override def controlsConnectivity = tileEntity.controlsConnectivity
  }

  class DefaultImpl extends Colored {
    var color = 0

    override def getColor = color

    override def setColor(value: Int): Unit = color = value

    override def controlsConnectivity = false
  }

  class DefaultStorage extends Capability.IStorage[Colored] {
    override def writeNBT(capability: Capability[Colored], t: Colored, Direction: Direction): INBT = {
      val color = t.getColor
      IntNBT.valueOf(color)
    }

    override def readNBT(capability: Capability[Colored], t: Colored, Direction: Direction, nbtBase: INBT): Unit = {
      nbtBase match {
        case nbt: IntNBT =>
          t.setColor(nbt.getAsInt)
        case _ =>
      }
    }
  }

}

File: src/main/scala/li/cil/oc/common/capabilities/CapabilityColored.java
package li.cil.oc.common.capabilities;

import li.cil.oc.api.tileentity.Colored;
import net.minecraft.nbt.NBTBase;
import net.minecraft.nbt.NBTTagInt;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;
import net.minecraftforge.common.capabilities.CapabilityManager;

import javax.annotation.Nullable;

public final class CapabilityColored {
    @CapabilityInject(Colored.class)
    public static Capability<Colored> COLORED_CAPABILITY;

    public static void register() {
        CapabilityManager.INSTANCE.register(Colored.class, new Capability.IStorage<Colored>() {
            @Nullable
            @Override
            public NBTBase writeNBT(final Capability<Colored> capability, final Colored instance, final EnumFacing side) {
                return instance.serializeNBT();
            }

            @Override
            public void readNBT(final Capability<Colored> capability, final Colored instance, final EnumFacing side, final NBTBase nbt) {
                if (nbt instanceof NBTTagInt) {
                    instance.deserializeNBT((NBTTagInt) nbt);
                }
            }
        }, () -> new Colored() {
            @Override
            public int getColor() {
                return 0;
            }

            @Override
            public boolean setColor(final int value) {
                return false;
            }

            @Override
            public boolean consumesDye() {
                return false;
            }

            @Override
            public boolean controlsConnectivity() {
                return false;
            }

            @Override
            public NBTTagInt serializeNBT() {
                return new NBTTagInt(0);
            }

            @Override
            public void deserializeNBT(final NBTTagInt nbt) {
            }
        });
    }

    private CapabilityColored() {
    }
}
